src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java

```diff
@@ -256,43 +270,74 @@ private void fileAclOp(AclOp op, Dataset dataset, List<String> fileIds, Map<IamR
-                    bucketForFile = dataLocationService.lookupBucket(file.getBucketResourceId());
-                    bucketCache.put(file.getBucketResourceId(), bucketForFile);
-                }
-                Storage storage = storageForBucket(bucketForFile);
-                URI gsUri = URI.create(file.getGspath());
-                String bucketPath = StringUtils.removeStart(gsUri.getPath(), "/");
-                BlobId blobId = BlobId.of(bucketForFile.getName(), bucketPath);
-                switch (op) {
-                    case ACL_OP_CREATE:
-                        for (Acl acl : acls) {
-                            storage.createAcl(blobId, acl);
-                        }
-                        break;
-                    case ACL_OP_DELETE:
-                        for (Acl.Group group : groups) {
-                            storage.deleteAcl(blobId, group);
-                        }
-                        break;
-                }
+
+            try (Stream<FSFile> stream = files.stream()) {
+                List<Future<FSFile>> futures = stream
+                    .map(file -> executor.submit(performAclCommand(bucketCache, file, op, acls, groups)))
+                    .collect(Collectors.toList());
+
+                FutureUtils.waitFor(futures);
             }
         }
+
+        performanceLogger.timerEndAndLog(
+            retrieveTimer,
+            dataset.getId().toString(), // not a flight, so no job id
+            this.getClass().getName(),
+            "gcsPdao.performAclCommands");
+    }
+
+    /**
+     * Perform the ACL setting commands on a specific file.
+     */
+    private Callable<FSFile> performAclCommand(final Map<String, GoogleBucketResource> bucketCache,
+                                               final FSFile file,
+                                               final AclOp op,
+                                               final List<Acl> acls,
+                                               final List<Acl.Group> groups) {
+        return () -> {
+            // Cache the bucket resources to avoid repeated database lookups.
+            // Synchronizing this block since this gets called with a potentially high degree of concurrency.
+            // Minimal overhead to lock here since 99% of the time this will be a simple map lookup
+            final GoogleBucketResource bucketForFile;
+            synchronized (bucketCache) {
+                bucketForFile = bucketCache.computeIfAbsent(
+                    file.getBucketResourceId(),
+                    k -> dataLocationService.lookupBucket(file.getBucketResourceId())
+                );
+            }
+            final Storage storage = storageForBucket(bucketForFile);
+            final URI gsUri = URI.create(file.getGspath());
+            final String bucketPath = StringUtils.removeStart(gsUri.getPath(), "/");
+            final BlobId blobId = BlobId.of(bucketForFile.getName(), bucketPath);
+            switch (op) {
```

Comment:
Whenever I see a `switch` over an `enum` like this, I think about rewriting it using enum methods. This avoids the need for `default` and makes the calling code cleaner. I wrote it up like this:

```
    private enum AclOp {
        ACL_OP_CREATE {
            void handle(Storage storage, BlobId blobId, List<Acl> acls, List<Acl.Group> groups) {
                for (Acl acl : acls) {
                    storage.createAcl(blobId, acl);
                }
            }
        },
        ACL_OP_DELETE {
            void handle(Storage storage, BlobId blobId, List<Acl> acls, List<Acl.Group> groups) {
                for (Acl.Group group : groups) {
                    storage.deleteAcl(blobId, group);
                }
            }
        };

        abstract void handle(Storage storage, BlobId blobId, List<Acl> acls, List<Acl.Group> groups);
    }
```
but maybe it would read better if the rest was moved into `AclOp` as well.